mitiq/shadows/shadows.py

```diff
@@ -0,0 +1,199 @@
+)
+from mitiq.shadows.quantum_processing import random_pauli_measurement
+
+
+def pauli_twirling_calibrate(
+    k_calibration: int = 1,
+    locality: Optional[int] = None,
+    zero_state_shadow_outcomes: Optional[Tuple[List[str], List[str]]] = None,
+    qubits: Optional[List[cirq.Qid]] = None,
+    executor: Optional[Callable[[cirq.Circuit], MeasurementResult]] = None,
+    num_total_measurements_calibration: Optional[int] = 20000,
+) -> Dict[str, complex]:
+    r"""
+    This function returns the dictionary of the median of means estimation
+    of Pauli fidelities: :math:`\{`"b": :math:`f_{b}\}_{b\in\{0,1\}^n}`.
+    The number of :math:`f_b` is :math:`2^n`, or :math:`\sum_{i=1}^d C_n^i` if
+    the locality :math:`d` is given.
+
+    In the notation of arXiv:2011.09636, this function estimates the
+    coefficient :math:`f_b`, which are expansion coefficients of the twirled
+    channel :math:`\mathcal{M}=\sum_b f_b\Pi_b`.
+
+    Args:
+        k_calibration: Number of groups of "median of means" used to solve for
+            Pauli fidelity.
+        locality: The locality of the operator, whose expectation value is
+            going to be estimated by the classical shadow. e.g. if operator is
+            Ising model Hamiltonian with nearist neighbour interaction, then
+            locality = 2.
+        zero_state_shadow_outcomes: The output of function
+            :func:`shadow_quantum_processing` of zero calibrate state.
+        qubits: The qubits to measure, needs to specify when the
+            `zero_state_shadow_outcomes` is None.
+        executor: The function to use to do quantum measurement, must be same
+            as executor in `shadow_quantum_processing`. Needs to specify when
+            the `zero_state_shadow_outcomes` is None.
+        num_total_measurements_calibration: Number of shots per group of
+            "median of means" used for calibration. Needs to specify when
+            the `zero_state_shadow_outcomes` is None.
+
+    Returns:
+        A dictionary containing the calibration outcomes.
+    """
+    if zero_state_shadow_outcomes is None:
+        if qubits is None:
+            raise ValueError("qubits should not be None.")
+        if executor is None:
+            raise ValueError("executor should not be None.")
+        if num_total_measurements_calibration is None:
+            raise ValueError(
+                "num_total_measurements_calibration should not be None."
+            )
+
+        # calibration circuit is of same qubit number with original circuit
+        zero_circuit = cirq.Circuit()
+        # perform random Pauli measurement one the calibration circuit
+        calibration_measurement_outcomes = random_pauli_measurement(
+            zero_circuit,
+            n_total_measurements=num_total_measurements_calibration,
+            executor=executor,
+            qubits=qubits,
+        )
+    else:
+        calibration_measurement_outcomes = zero_state_shadow_outcomes
```

Comment:
according to codecove, this case (`zero_state_shadow_outcomes` note None) is not tested. Could you test it somewhere in the test file?

Reply:
sure, will be add in the next version!